core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java

```diff
@@ -1464,14 +1589,29 @@ public Command workerHeartbeat(long workerId, Map<String, Long> capacityBytesOnT
       Map<BlockLocation, List<Long>> addedBlocks,
       Map<String, StorageList> lostStorage,
       List<Metric> metrics) {
-    if (isNotServing(workerId)) {
-      return Command.newBuilder().setCommandType(CommandType.Nothing).build();
-    }
     MasterWorkerInfo worker = mWorkers.getFirstByField(ID_INDEX, workerId);
     if (worker == null) {
+      /*
+       * If the worker is not recognized:
+       * 1. The worker never registered to the cluster, or the master has restarted/failover
+       * 2. The worker has been decommissioned and removed from the active worker list
+       */
+      worker = mDecommissionedWorkers.getFirstByField(ID_INDEX, workerId);
+      if (worker != null) {
+        if (isRejected(worker.getWorkerAddress())) {
+          throw new UnavailableRuntimeException(String.format(WORKER_DISABLED,
```

Comment:
why throw an exception instead of returning  a `Command.NOTHING`?

Reply:
return a Command.DISABLED